File: app/src/main/java/com/example/opengl/Texture.java
package com.example.opengl;

import static android.opengl.GLES20.GL_CLAMP_TO_EDGE;
import static android.opengl.GLES20.GL_LINEAR;
import static android.opengl.GLES20.GL_RGBA;
import static android.opengl.GLES20.GL_TEXTURE0;
import static android.opengl.GLES20.GL_TEXTURE_2D;
import static android.opengl.GLES20.GL_TEXTURE_MAG_FILTER;
import static android.opengl.GLES20.GL_TEXTURE_MIN_FILTER;
import static android.opengl.GLES20.GL_TEXTURE_WRAP_S;
import static android.opengl.GLES20.GL_TEXTURE_WRAP_T;
import static android.opengl.GLES20.glActiveTexture;
import static android.opengl.GLES20.glBindTexture;
import static android.opengl.GLES20.glDeleteTextures;
import static android.opengl.GLES20.glGenTextures;
import static android.opengl.GLES20.glGenerateMipmap;
import static android.opengl.GLES20.glTexParameteri;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.opengl.GLUtils;

public class Texture {
    public final int handle;
    public final int tilesCount;

    public Texture(int handle, int tilesCount) {
        this.handle = handle;
        this.tilesCount = tilesCount;
    }

    public static Texture Create(Context context, int resourceId, int tilesCount) {
        // Generate handle
        final int[] textureIds = new int[1];
        glGenTextures(1, textureIds, 0);
        int handle = textureIds[0];
        if (textureIds[0] == 0) {
            return null;
        }

        // Bind the handle
        glActiveTexture(GL_TEXTURE0);
        glBindTexture(GL_TEXTURE_2D, handle);

        // create Bitmap
        final BitmapFactory.Options options = new BitmapFactory.Options();
        options.inScaled = false;
        options.inPreferredConfig = Bitmap.Config.ARGB_8888;

        final Bitmap bitmap = BitmapFactory.decodeResource(
                context.getResources(), resourceId, options);

        if (bitmap == null) {
            glDeleteTextures(1, textureIds, 0);
            return null;
        }

        //GL.TexImage2D(TextureTarget.Texture2D, 0, PixelInternalFormat.Rgba, palette.Count, 1, 0, PixelFormat.Rgba, PixelType.UnsignedByte, palette.ToArray());
        //GLUtils.texImage2D(GL_TEXTURE_2D, 0, bitmap, 0);
        
        GLUtils.texImage2D(GL_TEXTURE_2D, 0, GL_RGBA, bitmap, 0);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);

        // Generate mipmaps.
        glGenerateMipmap(GL_TEXTURE_2D);

        bitmap.recycle();

        // reset target
        glBindTexture(GL_TEXTURE_2D, 0);

        return new Texture(handle, tilesCount);
    }

    // Activate texture
    // Multiple textures can be bound, if your shader needs more than just one.
    // If you want to do that, use GL.ActiveTexture to set which slot GL.BindTexture binds to.
    // The OpenGL standard requires that there be at least 16, but there can be more depending on your graphics card.
    public void Use(int unit) {
        glActiveTexture(unit);
        glBindTexture(GL_TEXTURE_2D, handle);
    }

    protected void finalize() {
        if (handle > 0) {
            final int[] textureIds = new int[1];
            textureIds[0] = handle;
            glDeleteTextures(1, textureIds, 0);
        }
    }
}

File: app/src/main/java/com/example/opengl/Texture.kt
package com.example.opengl

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.opengl.GLES20
import android.opengl.GLUtils

class Texture(private val handle: Int, val tilesCount: Int,  val width: Int, val height: Int) {
    fun use(unit: Int) {
        GLES20.glActiveTexture(unit)
        GLES20.glBindTexture(GLES20.GL_TEXTURE_2D, handle)
    }

    protected fun finalize() {
        if (handle > 0) {
            val textureIds = IntArray(1)
            textureIds[0] = handle
            GLES20.glDeleteTextures(1, textureIds, 0)
        }
    }

    companion object {
        fun create(context: Context, resourceId: Int, tilesCount: Int): Texture? {
            // Generate handle
            val textureIds = IntArray(1)
            GLES20.glGenTextures(1, textureIds, 0)
            val handle = textureIds[0]
            if (textureIds[0] == 0) {
                return null
            }

            // Bind the handle
            GLES20.glActiveTexture(GLES20.GL_TEXTURE0)
            GLES20.glBindTexture(GLES20.GL_TEXTURE_2D, handle)

            // create Bitmap
            val options = BitmapFactory.Options()
            options.inScaled = false
            options.inPreferredConfig = Bitmap.Config.ARGB_8888
            val bitmap = BitmapFactory.decodeResource(
                context.resources, resourceId, options
            )
            if (bitmap == null) {
                GLES20.glDeleteTextures(1, textureIds, 0)
                return null
            }
            val width = bitmap.width
            val height = bitmap.height

            //GL.TexImage2D(TextureTarget.Texture2D, 0, PixelInternalFormat.Rgba, palette.Count, 1, 0, PixelFormat.Rgba, PixelType.UnsignedByte, palette.ToArray());
            //GLUtils.texImage2D(GL_TEXTURE_2D, 0, bitmap, 0);
            GLUtils.texImage2D(GLES20.GL_TEXTURE_2D, 0, GLES20.GL_RGBA, bitmap, 0)
            GLES20.glTexParameteri(
                GLES20.GL_TEXTURE_2D,
                GLES20.GL_TEXTURE_MIN_FILTER,
                GLES20.GL_LINEAR
            )
            GLES20.glTexParameteri(
                GLES20.GL_TEXTURE_2D,
                GLES20.GL_TEXTURE_MAG_FILTER,
                GLES20.GL_LINEAR
            )
            GLES20.glTexParameteri(
                GLES20.GL_TEXTURE_2D,
                GLES20.GL_TEXTURE_WRAP_S,
                GLES20.GL_CLAMP_TO_EDGE
            )
            GLES20.glTexParameteri(
                GLES20.GL_TEXTURE_2D,
                GLES20.GL_TEXTURE_WRAP_T,
                GLES20.GL_CLAMP_TO_EDGE
            )

            // Generate mipmaps.
            GLES20.glGenerateMipmap(GLES20.GL_TEXTURE_2D)
            bitmap.recycle()

            // reset target
            GLES20.glBindTexture(GLES20.GL_TEXTURE_2D, 0)
            return Texture(handle, tilesCount, width, height)
        }

        fun create(context: Context, bitmap: Bitmap): Texture? {
            // Generate handle
            val textureIds = IntArray(1)
            GLES20.glGenTextures(1, textureIds, 0)
            val handle = textureIds[0]
            if (textureIds[0] == 0) {
                return null
            }

            // Bind the handle
            GLES20.glActiveTexture(GLES20.GL_TEXTURE0)
            GLES20.glBindTexture(GLES20.GL_TEXTURE_2D, handle)
            val width = bitmap.width
            val height = bitmap.height

            // set 1 byte pixel alignment
            GLES20.glPixelStorei(GLES20.GL_UNPACK_ALIGNMENT, 1)
            GLUtils.texImage2D(GLES20.GL_TEXTURE_2D, 0, bitmap,0)
            GLES20.glTexParameteri(
                GLES20.GL_TEXTURE_2D,
                GLES20.GL_TEXTURE_MIN_FILTER,
                GLES20.GL_LINEAR
            )
            GLES20.glTexParameteri(
                GLES20.GL_TEXTURE_2D,
                GLES20.GL_TEXTURE_MAG_FILTER,
                GLES20.GL_LINEAR
            )
            GLES20.glTexParameteri(
                GLES20.GL_TEXTURE_2D,
                GLES20.GL_TEXTURE_WRAP_S,
                GLES20.GL_CLAMP_TO_EDGE
            )
            GLES20.glTexParameteri(
                GLES20.GL_TEXTURE_2D,
                GLES20.GL_TEXTURE_WRAP_T,
                GLES20.GL_CLAMP_TO_EDGE
            )

            GLES20.glGenerateMipmap(GLES20.GL_TEXTURE_2D)

            GLES20.glBindTexture(GLES20.GL_TEXTURE_2D, 0)
            // set default (4 byte) pixel alignment
            GLES20.glPixelStorei(GLES20.GL_UNPACK_ALIGNMENT, 4)
            return Texture(handle, 0, width, height)
        }
    }
}